async-nats/src/client.rs

```diff
@@ -207,39 +211,71 @@ impl Client {
     }
 
     pub async fn request(&self, subject: String, payload: Bytes) -> Result<Message, Error> {
-        let inbox = self.new_inbox();
-        let mut sub = self.subscribe(inbox.clone()).await?;
-        self.publish_with_reply(subject, inbox, payload).await?;
-        self.flush().await?;
-        match sub.next().await {
-            Some(message) => {
-                if message.status == Some(StatusCode::NO_RESPONDERS) {
-                    return Err(Box::new(std::io::Error::new(
-                        ErrorKind::NotFound,
-                        "nats: no responders",
-                    )));
-                }
-                Ok(message)
-            }
-            None => Err(Box::new(io::Error::new(
-                ErrorKind::BrokenPipe,
-                "did not receive any message",
-            ))),
-        }
+        let request = RequestBuilder::new().payload(payload);
+        self.send_request(subject, request).await
     }
 
+    /// Sends the request with headers.
+    ///
+    /// # Examples
+    /// ```no_run
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<(), async_nats::Error> {
+    /// let client = async_nats::connect("demo.nats.io").await?;
+    /// let mut headers = async_nats::HeaderMap::new();
+    /// headers.insert("Key", "Value");
+    /// client.request_with_headers("service".into(), headers, "data".into()).await?;
+    /// # Ok(())
+    /// # }
+    /// ```
     pub async fn request_with_headers(
         &self,
         subject: String,
         headers: HeaderMap,
         payload: Bytes,
     ) -> Result<Message, Error> {
-        let inbox = self.new_inbox();
+        let request = RequestBuilder::new().headers(headers).payload(payload);
+        self.send_request(subject, request).await
+    }
+
+    /// Sends the request created by the [RequestBuilder].
+    ///
+    /// # Examples
+    /// ```no_run
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<(), async_nats::Error> {
+    /// let client = async_nats::connect("demo.nats.io").await?;
+    /// let request = async_nats::RequestBuilder::new().payload("data".into());
+    /// client.send_request("service".into(), request).await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn send_request(
```

Comment:
Hmm, just thinking out loud here

```suggestion
    pub async fn publish_request
```